Add offline behavior and error handling for `AcceptWalletTerms`

francoisl opened the pull request:
On hold for https://github.com/Expensify/Web-Expensify/pull/34765, but ready to review

### Details
- Add the red brick road and offline behavior for:
     - [x] Transfer balance
     - [x] Payment request
- On the final step of the KYC flow (ActiveStep), change the button text to either "Continue to payment" or "Continue to transfer" based on what action the user initially took to get there

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/218502

### Tests
Pro-tip: you can rerun the four queries below any time to reset your wallet to the Terms step and try the different test steps.

- Initial setup
1. Create a new account, get its wallet's `bankAccountID`
2. Run the following queries to simulate passing ExpectID and Onfido tests, and set the current KYC step to the Terms step (replace XXX with the wallet's ID)
```sql
UPDATE bankAccounts SET additionalData = JSON_SET(additionalData, '$.currentStep', 'TermsStep') WHERE bankAccountID = XXX;
UPDATE bankAccounts SET additionalData = JSON_SET(additionalData, '$.hasAcceptedTerms', JSON('false')) WHERE bankAccountID = XXX;
UPDATE bankAccounts SET additionalData = JSON_SET(additionalData, '$.verifications.externalApiResponses.requestorIdentityOnfido.status', 'pass') WHERE bankAccountID = XXX;
UPDATE bankAccounts SET additionalData = JSON_SET(additionalData, '$.verifications.externalApiResponses.requestorIdentityID.apiResult.id-number', 'IDO1234567891', '$.accountHistory', JSON_ARRAY(JSON_OBJECT('statusChangeTo', 'Open', 'statusChangeDate', '1623962658000000'))) WHERE bankAccountID = XXX;
UPDATE bankAccounts set addressState = false WHERE bankAccountID = XXX;
```

#### Transfer Balance flow

1. From an account `A` with a Gold wallet, send money to an account `B` with a non-validated wallet
2. In account B, go to Settings > Payments > Transfer Balance
3. Comment out [these lines](https://github.com/Expensify/Web-Expensify/blob/f5091aecdd331211e11600e859b0977252bb036d/lib/BankAccountAPI.php#L1480-L1492) to simulate not receiving the wallet activation response from the backend
4. Complete the KYC flow, submit
5. Make sure you see the "Wallet activated" page, with a button "Continue to transfer"
<img width="200" alt="image" src="https://user-images.githubusercontent.com/2229301/188103465-2eaa653f-43ab-4d6d-b549-3d16817a0302.png">
6. Go back to the Transfer balance page
7. In the Chrome console > Network, set your network as "Offline"
8. Make sure the balance amount gets grayed out
<img width="200" alt="image" src="https://user-images.githubusercontent.com/2229301/188104242-ef720283-1c98-44ac-85f5-eb26ee56c8f4.png">
7. Go back online, run the following in the JS console to make Onyx think you haven't tried to validate your wallet yet:

```js
Onyx.merge('userWallet', {shouldShowWalletActivationSuccess:null});
```

8. Invert [this line](https://github.com/Expensify/Web-Expensify/blob/f5091aecdd331211e11600e859b0977252bb036d/lib/BankAccountAPI.php#L1448): `if ($hasAcceptedTerms) {` to simulate the backend running into an error when you try to validate the bank account
9. Try the transfer money flow again
10. Make sure you see a red brick road leading to the transfer page after the backend sends the error push updates

https://user-images.githubusercontent.com/2229301/188111512-e0c37207-e24a-40a6-be37-7ca64468d64a.mov

11. Click on the "x" next to the error on the available balance view, make sure the error disappears


#### IOU / Payment flow

1. Change [this line](https://github.com/Expensify/Web-Expensify/blob/f5091aecdd331211e11600e859b0977252bb036d/lib/BankAccountAPI.php#L1448) to `if (!$hasAcceptedTerms) {`
1. From an account `A` with a Gold wallet, request money from an account `B` with a non-validated wallet
2. In account B, try to pay the request
3. When you get the wallet activated step, make sure the button reads "Continue to payment"
4. After the backend sends the error update, make sure you see that red brick indicator in the LHN, and that the IOU preview is grayed out with an error

https://user-images.githubusercontent.com/2229301/188113989-c5b9392e-36eb-4821-a270-a51cf64f9b1c.mov

6. Make sure you can clear the error by clicking on the X sign in the IOU preview
7. Go offline, make sure the IOU preview gets grayed out
<img width="607" alt="image" src="https://user-images.githubusercontent.com/2229301/188114407-f46701d8-4eb9-409e-95ce-640f8ea5e037.png">
- [x] Verify that no errors appear in the JS console

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>


### QA steps

None. We'll test the entire KYC process internally once all the steps have been refactored.

francoisl committed:
Add new prop types, use `OfflineWithFeedback` on base payment page

francoisl committed:
Red brick road for WalletTerms errors

francoisl committed:
Support clearing RBR wallet term errors

francoisl committed:
Merge main, fix conflicts

francoisl committed:
Make 'Continue' button text depend on source action

MariaHCD commented on src/pages/EnablePayments/walletTermsPropTypes.js:
```diff
@@ -0,0 +1,7 @@
+import PropTypes from 'prop-types';
+
+/**  */
```
Let's add a comment about the purpose of this prop type/what data it contains

francoisl committed:
Fix conflict

MariaHCD commented on src/components/SettlementButton.js:
```diff
@@ -80,6 +80,7 @@ class SettlementButton extends React.Component {
                 addBankAccountRoute={this.props.addBankAccountRoute}
                 addDebitCardRoute={this.props.addDebitCardRoute}
                 isDisabled={this.props.network.isOffline}
+                sourceAction="iou"
```
NAB but maybe `iou` can be a constant

MariaHCD commented on src/pages/EnablePayments/ActivateStep.js:
```diff
@@ -35,6 +40,13 @@ class ActivateStep extends React.Component {
     renderGoldWalletActivationStep() {
+        let continueButtonText = this.props.translate('common.continue');
+        if (this.props.userWallet.sourceAction) {
+            continueButtonText = this.props.userWallet.sourceAction === 'transferBalance'
```
`transferBalance` can be a constant too

francoisl replied:
(Same question as with `sourceAction="iou"`)

MariaHCD commented on src/pages/EnablePayments/TermsStep.js:
```diff
@@ -17,20 +17,18 @@ import Text from '../../components/Text';
-        /** Any additional error message to show */
-        errors: PropTypes.objectOf(PropTypes.string),
-    }),
+    // eslint-disable-next-line react/no-unused-prop-types
+    walletTerms: PropTypes.objectOf(walletTermsPropTypes),
```
Is the walletTerms prop not being used in this component?

MariaHCD commented on src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js:
```diff
@@ -275,7 +275,14 @@ class BasePaymentsPage extends React.Component {
                             <View style={[styles.mv5]}>
-                                <CurrentWalletBalance />
+                                <OfflineWithFeedback
+                                    pendingAction="add"
```
```suggestion
                                    pendingAction={CONST.RED_BRICK_ROAD_PENDING_ACTION.ADD}
```

MariaHCD commented on src/pages/settings/InitialSettingsPage.js:
```diff
@@ -138,7 +143,7 @@ class InitialSettingsPage extends React.Component {
                 icon: Expensicons.Wallet,
                 action: () => { Navigation.navigate(ROUTES.SETTINGS_PAYMENTS); },
-                brickRoadIndicator: PaymentMethods.hasPaymentMethodError(this.props.bankAccountList, this.props.cardList) || !_.isEmpty(this.props.userWallet.errors) ? 'error' : null,
+                brickRoadIndicator: PaymentMethods.hasPaymentMethodError(this.props.bankAccountList, this.props.cardList) || !_.isEmpty(this.props.userWallet.errors) || !_.isEmpty(this.props.walletTerms.errors) ? 'error' : null,
```
The linter is complaining about this line being too long. Guess it's time to split it into multiple lines :)

MariaHCD reviewed (commented):
Looks great so far!

francoisl committed:
Fix conflicts

francoisl committed:
Review comments

francoisl committed:
Add `OfflineWithFeedback` for IOUPreview component

francoisl committed:
Replace `sourceAction` with IOU's parent chat ID

francoisl committed:
Clear errors in LHN

francoisl commented:
Pushed all the changes discussed, so this should be ready to review.
My internet is too wonky to upload screenshots/videos for the tests at the moment, I'll add those tomorrow morning.

francoisl committed:
Pull main, fix conflict

francoisl commented:
Still ready to review :) Just fixed a small conflict.

jasperhuangg commented on src/components/ReportActionItem/IOUPreview.js:
```diff
@@ -124,61 +131,71 @@ const IOUPreview = (props) => {
+                            }}
+                            errorRowStyles={[styles.mbn1]}
+                        >
+                            <View>
+                                <View style={styles.flexRow}>
+                                    <View style={styles.flex1}>
+                                        <View style={styles.flexRow}>
```
NAB since I know these aren't related to your changes but that sure is a whole lotta nested Views, I don't think those are all necessary

jasperhuangg commented on src/pages/EnablePayments/TermsStep.js:
```diff
@@ -17,20 +17,23 @@ import Text from '../../components/Text';
-        /** Any additional error message to show */
-        errors: PropTypes.objectOf(PropTypes.string),
-    }),
+    // eslint-disable-next-line react/no-unused-prop-types
+    walletTerms: PropTypes.objectOf(walletTermsPropTypes),
```
NAB since I'm sure there's good reason for this but why are we including this prop if it's not used in this component?

MariaHCD replied:
Looks like its used here: https://github.com/Expensify/App/blob/6c8328bdc872b9579f38b67180eaff16e2c99106/src/pages/EnablePayments/TermsStep.js#L71

But the linter can't understand that.

jasperhuangg replied:
ahh my bad, thanks for clarifying!

jasperhuangg commented on src/pages/EnablePayments/TermsStep.js:
```diff
@@ -143,5 +147,8 @@ export default compose(
         walletTerms: {
             key: ONYXKEYS.WALLET_TERMS,
         },
```
Same question here, this is never used, why include it?

jasperhuangg reviewed (commented):
Couple questions but overall looks really good!

MariaHCD commented on src/libs/actions/Wallet.js:
```diff
@@ -96,6 +96,15 @@ function setWalletShouldShowFailedKYC(shouldShowFailedKYC) {
+ * @param {Number} chatReportID
+ */
+function setKYCWallSourceChatReportID(chatReportID) {
+    Onyx.merge(ONYXKEYS.USER_WALLET, {chatReportID});
```
Do you think that we should set the chatReportID in the WALLET_TERMS onyx key instead? Since it's related to the terms step vs. the user's wallet.

MariaHCD reviewed (commented):
Looks great! Just one merge conflict to fix.

francoisl committed:
Pull main, fix conflict

francoisl committed:
PR comments

- Set chatReportID from IOU flow in Onyx's `WALLET_TERMS` instead of `USER_WALLET`
- Merge two nested `<View>` into one
- Fix a bunch of React prop type errors

francoisl committed:
Fix `availablaBalance` propType error

francoisl committed:
Remove unused import

francoisl committed:
Removed remaining unused imports

francoisl commented:
Ok, PR updated and all checks are green now.

MariaHCD commented on src/components/KYCWall/kycWallPropTypes.js:
```diff
@@ -21,14 +21,18 @@ const propTypes = {
 
     /** The user's wallet */
-    userWallet: PropTypes.objectOf(userWalletPropTypes),
+    userWallet: userWalletPropTypes,
```
Out of curiosity, what's the difference between `userWallet: PropTypes.objectOf(userWalletPropTypes)` vs `userWallet: userWalletPropTypes` 👀

francoisl replied:
I couldn't find a super clear explanation in the official docs, the best I could find was [this](https://stackoverflow.com/a/47254950/9329222) which seems to say that if you use both `objectOf` and `shape`, it means you expect your prop to always contain the exact structure you declare. In our case, we don't always have an `availableBalance` in the `userWallet` I imagine.

MariaHCD reviewed (dismissed):
LGTM! In terms of QA, I imagine this will be pretty tricky to test because we won't have a way to simulate an error with AcceptWalletTerms on staging/prod 🤔

francoisl committed:
Don't show avatar indicator for IOU terms errors

francoisl commented on src/components/AvatarWithIndicator.js:
```diff
@@ -69,6 +74,7 @@ const AvatarWithIndicator = (props) => {
     // we only care if a single error exists anywhere.
     const errorCheckingMethods = [
         () => !_.isEmpty(props.userWallet.errors),
+        () => !_.isEmpty(props.walletTerms.errors),
```
Updated 

<table>
<tr>
    <td>IOU error</td>
    <td>Transfer error</td>
</tr></tr>
    <td><img width="940" alt="image" src="https://user-images.githubusercontent.com/2229301/189315233-1a81c11e-d1fc-4654-b3ce-6964e88188b4.png"></td>
    <td><img width="801" alt="image" src="https://user-images.githubusercontent.com/2229301/189315205-4b6fc69f-b316-45ef-b9ec-72721f5c57cd.png"></td>
</tr>
</table>

MariaHCD commented:
#### PR Reviewer Checklist
- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

MariaHCD reviewed (approved):
LGTM! Thanks for the thorough work on this!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

MariaHCD commented:
Hmm, the only failing test was the Contributors+ checklist and that passed when I re-ran it before merging: https://github.com/Expensify/App/actions/runs/3021293498

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3043829133) to staging by @MariaHCD in version: 1.2.0-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅
